### ISSUE TITLE
Layered effect: `examples/layeredeffects/Buffers.fst` does not verify

### ISSUE THREAD

**Issue by W95Psp** [FStarLang/FStar#2051]
Hi,

The module `Buffers.fst` fails with `(Error 19) assertion failed; The solver found a (partial) counterexample, try to spell your proof in more detail or increase fuel/ifuel
` in the examples from `layeredeffects` directory, at the following lines:
https://github.com/FStarLang/FStar/blob/f675047186fd0e6f4b0f6dd41fc1838c759f8912/examples/layeredeffects/Buffers.fst#L109-L118

Admitting SMT queries makes it work, but I think I recall this example file to verify properly at some point in the past. I tried a few F* versions (today's master, a few weeks ago, and an older one), but none were working for me.

I'm working on a layered effect that has the same flaw, for now I'll just admit locally SMT queries, but I would be happy to understand why Z3 is unhappy there ^^

**Comment by aseemr** [comment]
Thanks for the report @W95Psp. This failure is because at some point we changed the shape of the `return` repr to be `a:Type -> x:a -> <some binders> -> repr a is` where `is` are the effect indices. Earlier the `<some binders>` used to appear before `x:a`. In this example, if we change the `return` definition to be:

```
inline_for_extraction
let return (a:Type) (x:a) (b1 b2:B.buffer u8)  //note b1 and b2 are after the (x:a) parameter
: repr a b1 b2 (λ p s → p x s)
= λ _ → x
```

Then at leat the layered effect definition typechecks. However, there are still some failures later in the module. I removed this example from the regressions because the setup of having `lseq` in the specs was not working well.

If this is your exact scenario, I can try to refresh my memory about the issue and spend some time fixing it in the example. Please let me know and I will be happy to take a look, thanks!

**Comment by W95Psp** [comment]
Oh nice, thanks! I should have looked at another layered effect then!

Well, I'm dealing with buffers of constant length, but I indeed use a `Seq.index (B.as_seq somememory somebuffer) 0`. Was the problematic part the "keeping implicit lengths around" thing?
If it's not I'm fine I guess; I just keep a few constant-length buffers around plus a region. 

Anyway, thanks for your answer! :)

**Comment by aseemr** [comment]
Yes, it is mainly having to specify these implicits. If that's not an issue for you, the style should be fine. I will also try to restore this file some time soon.
